File: backend/ai/chat_stat_format.py
"""Shared number formatting for copilot chat stats (text + voice).

Keeps percentages, ETH amounts, and other decimals readable — never more than
three digits after the decimal point unless the value is a whole number.
"""
from __future__ import annotations

import math
import re
from typing import Any

CHAT_STAT_MAX_DECIMALS = 3


def format_chat_stat_number(
    raw: Any,
    *,
    max_decimals: int = CHAT_STAT_MAX_DECIMALS,
) -> str:
    """Format a numeric stat for chat display (no unit suffix)."""
    text = str(raw or "0").strip().replace("%", "").replace(",", "")
    if not text:
        return "0"
    try:
        value = float(text)
    except (TypeError, ValueError):
        return str(raw).strip() if raw not in (None, "") else "0"
    if not math.isfinite(value):
        return "0"
    if value == int(value):
        return str(int(value))
    places = max(0, int(max_decimals))
    formatted = f"{value:.{places}f}".rstrip("0").rstrip(".")
    return formatted or "0"


def format_chat_stat_percentage(
    raw: Any,
    *,
    max_decimals: int = CHAT_STAT_MAX_DECIMALS,
) -> str:
    """Percentage value without a trailing % sign."""
    return format_chat_stat_number(raw, max_decimals=max_decimals)


def format_chat_stat_eth_amount(
    raw: Any,
    *,
    max_decimals: int = CHAT_STAT_MAX_DECIMALS,
) -> str:
    """ETH amount for inline chat stats (no 'ETH' suffix)."""
    return format_chat_stat_number(raw, max_decimals=max_decimals)


def format_chat_stat_percentage_label(
    raw: Any,
    *,
    max_decimals: int = CHAT_STAT_MAX_DECIMALS,
) -> str:
    """Percentage with % suffix, e.g. '8.333%'."""
    return f"{format_chat_stat_percentage(raw, max_decimals=max_decimals)}%"


_LONG_DECIMAL_RE = re.compile(r"(?<![\d.])(\d+\.\d{4,})(?![\d])")


def normalize_chat_stat_text(
    text: str,
    *,
    max_decimals: int = CHAT_STAT_MAX_DECIMALS,
) -> str:
    """Normalize any long decimal literals embedded in a chat line."""
    if not text:
        return ""

    def _replace(match: re.Match[str]) -> str:
        return format_chat_stat_number(match.group(1), max_decimals=max_decimals)

    return _LONG_DECIMAL_RE.sub(_replace, text)

File: frontend/lib/chat-stat-format.ts
import { formatEthText } from "@/lib/utils";

export const CHAT_STAT_MAX_DECIMALS = 3;

/** Format a numeric chat stat with at most three decimal places. */
export function formatChatStatNumber(
  value: number | string | null | undefined,
  maxDecimals: number = CHAT_STAT_MAX_DECIMALS,
): string {
  const raw = String(value ?? "0").trim().replace(/%/g, "").replace(/,/g, "");
  if (!raw) return "0";
  const parsed = Number(raw);
  if (!Number.isFinite(parsed)) return raw;
  if (parsed === Math.trunc(parsed)) return String(Math.trunc(parsed));
  const fixed = parsed.toFixed(Math.max(0, maxDecimals));
  return fixed.replace(/\.?0+$/, "") || "0";
}

const LONG_DECIMAL_RE = /(?<![\d.])(\d+\.\d{4,})(?![\d])/g;

/** Normalize copilot chat text — trim clumsy long decimals in stats lines. */
export function formatChatStatText(
  text: string,
  maxDecimals: number = CHAT_STAT_MAX_DECIMALS,
): string {
  if (!text) return "";
  const withEth = formatEthText(text);
  return withEth.replace(LONG_DECIMAL_RE, (match) => formatChatStatNumber(match, maxDecimals));
}
